### DISCUSSION TITLE
What do we need to provide for the the upper layer?

### DISCUSSION THREAD

**Discussion by noneback** [casbin-mesh/neo#9] [Ideas]
Mainly in two aspects, **Service** and **API**

**Comment by noneback** [comment]
The API can be derived from Casbin Adapter Interface. 

We need to provide those services and APIs:
- Basic policies CURD.
- Build index on some attributes.

Underneath the data access methods we need to:
- Indexing on curtain attribute.
- Workload Optimizer. Predict pushdown or any other strategies.
- Transaction if necessary.
- Disk Management. Include file organization, page layout, and so on.
- Buffer Pool Management. Manage data loaded from disk.

**Reply by WenyXu** [reply]
I want to mention one thing, the policies of Casbin can be classified into two types generally(typical in models like RBAC):
- One stands for **policies of ability**, such as `p Alice data1 read `, which means Alice can read data1
- Another one stands for **user groups or resources groups**, such as `g Alice admin,` which means Alice is the admin member.

For more details: 
https://casbin.org/docs/en/supported-models
https://casbin.org/docs/en/rbac

**Comment by noneback** [comment]
We can design policies storage engine focusing on the skeleton above. We can add more details.
Any advice to modify or update the skeleton?
